hbase-server/src/main/java/org/apache/hadoop/hbase/master/MasterRpcServices.java

```diff
@@ -2465,6 +2466,39 @@ public GetTableStateResponse setTableStateInMeta(RpcController controller,
     }
   }
 
+  /**
+   * Update state of the region in meta only. This is required by hbck in some situations to cleanup
+   * stuck assign/ unassign regions procedures for the table.
+   *
+   * @return previous state of the region
+   */
+  @Override
+  public MasterProtos.GetRegionStateResponse setRegionStateInMeta(RpcController controller,
+    MasterProtos.SetRegionStateInMetaRequest request) throws ServiceException {
+    try {
+      RegionInfo info = this.master.getAssignmentManager().
+        loadRegionFromMeta(request.getRegionInfo().getRegionEncodedName());
+      LOG.trace("region info loaded from meta table: {}", info);
+      RegionState prevState = this.master.getAssignmentManager().getRegionStates().
+        getRegionState(info);
+      RegionState newState = RegionState.convert(request.getRegionState());
+      LOG.info("{} set region={} state from {} to {}", master.getClientIdAuditPrefix(),
+        info, prevState.getState(), newState.getState());
+      Put metaPut = MetaTableAccessor.makePutFromRegionInfo(info, System.currentTimeMillis());
+      metaPut.addColumn(HConstants.CATALOG_FAMILY,
+        HConstants.STATE_QUALIFIER, Bytes.toBytes(newState.getState().name()));
+      List<Put> putList = new ArrayList<>();
+      putList.add(metaPut);
+      MetaTableAccessor.putsToMetaTable(this.master.getConnection(), putList);
+      //Loads from meta again to refresh AM cache with the new region state
+      this.master.getAssignmentManager().loadRegionFromMeta(info.getEncodedName());
+      return MasterProtos.GetRegionStateResponse.newBuilder().
+        setRegionState(prevState.convert()).build();
+    } catch (Exception e) {
+      throw new ServiceException(e);
+    }
```

Comment:
This is great but what you think about doing more than one Region per RPC?

I'm thinking of the case where you have a cluster with thousands of Regions and perhaps a few hundred need their state set .  If we only do a single Region at a time, it will take a long time settiing state on hundreds of Regions. We'd have to change the hbck2 command too so it took more than one Region in the list.

Reply:
Yeah, had thought about that after this commit. Let me refactor it to receive list of region encoded names.